includes/class-amp-theme-support.php

```diff
@@ -347,7 +347,14 @@ public static function finish_output_buffering( $output ) {
 			1
 		);
 
+		$dom = AMP_DOM_Utils::get_dom( $output );
+		// Sanitize forms in the document.
+		$sanitizer = new AMP_Form_Sanitizer( $dom );
+		$sanitizer->sanitize();
+
 		// @todo Add more validation checking and potentially the whitelist sanitizer.
+		$output = $dom->saveHTML();
```

Comment:
@DavidCramer I chatted about this with @ThierryA, but I think we should avoid having to load the entire response into the DOM for sanitization. I think that we should limit sanitization to just the post content, 3rd-party widgets, and other “leaf nodes” which we know are liable to be invalid AMP.

Is this right? Or am I off track? My worry is that loading the HTML into a DOMDocument for every request will be too heavy. But at the same time, if we're already spinning up a DOMDocument for every instance of `the_content()` maybe it is actually better to just do one for the entire HTML. This would certainly simplify somethings for sanitizing widgets (cc @kienstra). It could also mean that we wouldn't have to have this ad hoc logic to search for elements requiring components: https://github.com/Automattic/amp-wp/blob/e60cb152a50e2ffbf5504d41aee859ad1d0e0baa/includes/class-amp-theme-support.php#L448-L455

Maybe we should decide to not prematurely optimize and instead keep things simple at first and just go ahead and sanitize the entire response. Thoughts?

Reply:
See findings at https://github.com/Automattic/amp-wp/issues/875#issuecomment-358491587

It seems we should definitely not sanitize the entire output buffer.

Reply:
@westonruter whats the consensus now on this? It looks like your findings indicated it may be a good idea to sanitize the whole buffer output. I personally think this is the most efficient. I spoke with @ThierryA about this yesterday and he was going to do some tests, like you did.

Reply:
I'm leaning toward sanitizing the entire output buffer now.

Reply:
@DavidCramer do you want to build off of https://github.com/Automattic/amp-wp/issues/875#issuecomment-358560145 in a new PR to serve as a basis for what you're introducing here with form sanitization?